### PR TITLE
Normalize CPU percentages when using cgroup v2

### PR DESCRIPTION
**Pull request by luismiramirez**
If the cgroup v2 `cpu.max` file contains a value other than `max` for its first segment, we're able to calculate the number of used CPUs in containers. If that's possible, the usage percentage is divided by the number of CPUs.

This change will result in a non-backward compatible change for those relying on > 100% old values.

Meaning that if there're 5 CPUs available and they're all at 100% usage, the usage won't be reported anymore as 500% but as 100%.

Part of: #67

### CODE REVIEW

File: src/cpu/cgroup.rs [hunk 1]
```diff
@@ -79,8 +79,9 @@ pub fn read() -> Result<CgroupCpuMeasurement> {
     use super::cgroup_v2::read_and_parse_v2_sys_stat;
 
     let v2_sys_fs_file = Path::new("/sys/fs/cgroup/cpu.stat");
+    let v2_sys_fs_cpu_max_file = Path::new("/sys/fs/cgroup/cpu.max");
     if v2_sys_fs_file.exists() {
```
Comment (tombruijn) [inline src/cpu/cgroup.rs:83]:
```suggestion
    if v2_sys_fs_file.exists() {
      let v2_sys_fs_cpu_max_file = Path::new("/sys/fs/cgroup/cpu.max");
```
No need to set the variable if the cgroups v2 CPU file doesn't exist.

File: src/cpu/cgroup_v2.rs [hunk 2]
```diff
@@ -64,29 +88,45 @@ mod test {
...
+    }
+
     #[test]
     fn test_read_v2_sys_wrong_path() {
-        match read_and_parse_v2_sys_stat(&Path::new("bananas")) {
+        match read_and_parse_v2_sys_stat(&Path::new("bananas"), &Path::new("potato")) {
             Err(ProbeError::IO(_, _)) => (),
             r => panic!("Unexpected result: {:?}", r),
         }
     }
```
Comment (tombruijn) [inline src/cpu/cgroup_v2.rs:122]:
```suggestion
    }
    
    #[test]
    fn test_read_v2_sys_wrong_path_for_max() {
        match read_and_parse_v2_sys_stat(&Path::new("fixtures/linux/sys/fs/cgroup_v2/cpu.stat_1"), &Path::new("potato")) {
            Err(ProbeError::IO(_, _)) => (),
            r => panic!("Unexpected result: {:?}", r),
        }
    }
```

I would add another test for when only the max file doesn't exist. Make sure it returns the expected error too.

File: src/cpu/cgroup_v2.rs [hunk 3]
```diff
@@ -95,30 +135,35 @@ mod test {
     #[test]
     fn test_read_and_parse_v2_sys_stat_garbage() {
         let path = Path::new("fixtures/linux/sys/fs/cgroup_v2/cpu.stat_garbage");
-        match read_and_parse_v2_sys_stat(&path) {
+        let max_file_path = Path::new("fixtures/linux/fs/cgroup_v2/cpu.max");
+        match read_and_parse_v2_sys_stat(&path, &max_file_path) {
             Err(ProbeError::UnexpectedContent(_)) => (),
             r => panic!("Unexpected result: {:?}", r),
         }
     }
```
Comment (tombruijn) [inline src/cpu/cgroup_v2.rs:143]:
```suggestion
    }
    

    #[test]
    fn test_read_and_parse_v2_sys_max_garbage() {
        let path = Path::new("fixtures/linux/sys/fs/cgroup_v2/cpu.stat_1");
        let max_file_path = Path::new("fixtures/linux/fs/cgroup_v2/cpu.max_garbage");
        match read_and_parse_v2_sys_stat(&path, &max_file_path) {
            Err(ProbeError::UnexpectedContent(_)) => (),
            r => panic!("Unexpected result: {:?}", r),
        }
    }
```

Could use a test for what happens when the max file is garbage. Would need a `cpu.max_garbage` fixture with content like `max garbage` in there.

File: src/cpu/cgroup_v2.rs (hunk 2 above)
Comment (luismiramirez) [reply src/cpu/cgroup_v2.rs:122]: Do we want it to fail if the max file doesn't exist? We can still provide valuable data if it's not present.

File: src/cpu/cgroup_v2.rs (hunk 3 above)
Comment (luismiramirez) [reply src/cpu/cgroup_v2.rs:143]: What do you mean?

If the first segment is `max`, it's already going to ignore it. Is it a possible scenario?

File: src/cpu/cgroup_v2.rs (hunk 2 above)
Comment (tombruijn) [reply src/cpu/cgroup_v2.rs:122]: Oh yeah it should fall back on the default behavior not return an error here.

**Review by tombruijn — APPROVED** [review]

File: src/cpu/cgroup_v2.rs [hunk 4]
```diff
@@ -7,9 +7,29 @@ use std::path::Path;
...
+    if cpu_max_path.exists() {
+        let reader = file_to_buf_reader(&cpu_max_path)?;
+        for line in reader.lines() {
+            let line = line.map_err(|e| ProbeError::IO(e, path_to_string(path)))?;
+            let segments: Vec<&str> = line.split_whitespace().collect();
+            let max = segments[0];
+            let period = parse_u64(&segments[1])?;
+
+            if max != "max" {
+                cpu_count = parse_u64(&max)? / period;
```
Comment (unflxw) [inline src/cpu/cgroup_v2.rs:27]: I think `cpu_count` will be an `u64`. This won't do the right thing when allocating fractional CPU capability, which is apparently something you can do:

```
> docker run --rm -it --cpus "0.5" ubuntu:22.04 bash
root@69aeb4ed7ca8:/# cat /sys/fs/cgroup/cpu.max
50000 100000
```

File: src/cpu/cgroup_v2.rs [hunk 5]
```diff
@@ -7,9 +7,29 @@ use std::path::Path;
...
+    path: &Path,
+    cpu_max_path: &Path,
+) -> Result<CgroupCpuMeasurement> {
+    // If the cpu.max file exists, we can use it to calculate the number of CPUs
+    // in the cgroup. It's also required that the first value is not set to "max",
+    // otherwise we can't calculate the number of CPUs.
+    let mut cpu_count = 0;
+    if cpu_max_path.exists() {
+        let reader = file_to_buf_reader(&cpu_max_path)?;
+        for line in reader.lines() {
```
Comment (unflxw) [inline src/cpu/cgroup_v2.rs:20]: Does it make sense to iterate through lines? We only expect exactly one line. Perhaps if there is more or less than one line we should treat it as a parsing error.

**Review by unflxw — APPROVED** [review]
Looks good modulo the comments. (I didn't leave them as part of the review because I have Monday brain)

File: src/cpu/cgroup_v2.rs (hunk 4 above)
Comment (luismiramirez) [reply src/cpu/cgroup_v2.rs:27]: I added two new test cases.

- `test_read_v2_sys_measurement_half_usage`
- `test_in_percentages_integration_v2_non_int_max`

The results show the behavior expected when there's a single CPU.

**Pull request merged**